RetailCoder.VBE/UI/DockableWindowHost.cs

```diff
@@ -130,37 +131,346 @@ protected override bool ProcessKeyPreview(ref Message m)
+            while (_userControl.Controls.Count > 0)
             {
-                Release();
+                var control = _userControl.Controls[0];
+                _userControl.Controls.Remove(control);
+                control.Dispose();
             }
-            base.DefWndProc(ref m);
         }
 
-        //override 
 
-        private bool _released; 
-        public void Release()
+        public int /* IOleObject:: */ SetClientSite([In] IntPtr /* IOleClientSite */ pClientSite)
         {
-            if (_released)
+            _logger.Log(LogLevel.Trace, "IOleObject::SetClientSite() called");
+
+            // IOleObject::SetClientSite is typically called with pClientSite = null just before calling IOleObject::Close()
+            // We release all host COM objects here in that instance.
+
+            _cachedClientSite?.Dispose();
+            _cachedClientSite = null;
+
+            if (pClientSite != IntPtr.Zero)
             {
-                return;
+                _cachedClientSite = new Wrapper_IOleClientSite(pClientSite);
+                return _userControl._IOleObject.SetClientSite(_cachedClientSite.PeekAggregatedReference());     // callee will take its own reference
             }
+            return (int)COMConstants.S_OK;
+        }
+
+        public int /* IOleObject:: */ GetClientSite([Out] out IntPtr /* IOleClientSite */ ppClientSite)
+        {
+            _logger.Log(LogLevel.Trace, "IOleObject::GetClientSite() called");
+            ppClientSite = _cachedClientSite?.CopyAggregatedReference() ?? IntPtr.Zero;
+            return (int)COMConstants.S_OK;
+        }
+
+        public int /* IOleObject:: */ SetHostNames([In, MarshalAs(UnmanagedType.LPWStr)] string szContainerApp, [In, MarshalAs(UnmanagedType.LPWStr)] string szContainerObj)
+        {
+            _logger.Log(LogLevel.Trace, "IOleObject::SetHostNames() called");
+            return _userControl._IOleObject.SetHostNames(szContainerApp, szContainerObj);
+        }
+
+        public int /* IOleObject:: */ Close([In] uint dwSaveOption)
+        {
+            _logger.Log(LogLevel.Trace, "IOleObject::Close() called");
+            int hr = _userControl._IOleObject.Close(dwSaveOption);
+
+            // IOleObject::SetClientSite is typically called with pClientSite = null just before calling IOleObject::Close()
+            // If it didn't, we release all host COM objects here instead,
+
+            // This is the point where we can deterministically, and safely release our COM references for this ActiveX control.
+            // Moreover, we can release the UserControl COM references, as Close() should be the very last call into the IOleObject interface.
+            PerformUserControlShutdown();
+
+            return hr;
+        }
 
-            Debug.WriteLine("DockableWindowHost release called.");
+        private void PerformUserControlShutdown()
+        {
+            ReleaseCOMReferenceOfSctiveXControl();
```

Comment:
Typo here - Also I think the naming should be `ReleaseComReferenceOfActiveXControl` --- it could be also shortened to `ReleaseActiveXControlReference`, possibly.